pynsot/commands/cmd_protocol_types.py

```diff
@@ -0,0 +1,300 @@
+)
+@click.option(
+    '-n',
+    '--name',
+    metavar='NAME',
+    type=str,
+    help='The name of the Protocol Type.',
+    required=True,
+)
+@click.option(
+    '-s',
+    '--site-id',
+    metavar='SITE_ID',
+    type=int,
+    help='Unique ID of the Site this Protocol Type is under.  [required]',
+    callback=callbacks.process_site_id,
+)
+@click.pass_context
+def add(ctx, required_attributes, description, name, site_id):
+    """
+    Add a new Protocol Type.
+
+    You must provide a Protocol Type name or ID using the UPDATE option.
+
+    When adding a new Protocol Type, you must provide a value for the -n/--name
+    option.
+
+    Examples: OSPF, BGP, etc.
+
+    You may also provide required Protocol attributes, you may specify the -r/--required_attributes
+    option once for each attribute. The Protocol attributes must exist before adding them to a
+    protocol type.
+
+    You must provide a Site ID using the -s/--site-id option.
+    """
+    data = ctx.params
+
+    # Remove if empty; allow default assignment
+    if description is None:
+        data.pop('description')
+
+    ctx.obj.add(data)
+
+
+# List
+@cli.group(invoke_without_command=True)
+@click.option(
+    '-r',
+    '--required_attributes',
+    metavar='ATTRIBUTE',
+    type=str,
+    help='The name of a Protocol attribute that will be used to filter the results.',
+    multiple=True,
+)
+@click.option(
+    '-d',
+    '--description',
+    metavar='DESCRIPTION',
+    type=str,
+    help='Filter by Protocol Type matching this description.',
+)
+@click.option(
+    '-i',
+    '--id',
+    metavar='ID',
+    type=types.NATURAL_KEY,
+    help='Unique ID of the Protocol Type being retrieved.',
+)
+@click.option(
+    '-n',
+    '--name',
+    metavar='NAME',
+    help='Filter to Protocol Type matching this name.'
+)
+@click.option(
+    '-s',
+    '--site-id',
+    metavar='SITE_ID',
+    help='Unique ID of the Site this Protocol Type is under.',
+    callback=callbacks.process_site_id,
+)
+@click.pass_context
+def list(ctx, required_attributes, description, id, name, site_id):
+    """
+    List existing Protocol Types for a Site.
+
+    You must provide a Site ID using the -s/--site-id option.
+
+    When listing Protocol Types, all objects are displayed by default. You
+    optionally may lookup a single Protocol Types by ID using the -i/--id option.
+    The ID can either be the numeric ID of the Protocol Type.
+    """
+    data = ctx.params
+
+    # If we provide ID, show more fields
+    if any([id, name]):
+        display_fields = VERBOSE_FIELDS
+    else:
+        display_fields = DISPLAY_FIELDS
+
+    # If we aren't passing a sub-command, just call list(), otherwise let it
+    # fallback to default behavior.
+    if ctx.invoked_subcommand is None:
+        ctx.obj.list(
+            data, display_fields=display_fields,
+            verbose_fields=VERBOSE_FIELDS
+        )
+
+@list.command()
+@click.pass_context
+def protocols(ctx, *args, **kwargs):
```

Comment:
There is not an API endpoint for this e.g. `/api/sites/1/protocol_types/1/protocols/`. This sub-command can be removed.

Reply:
You left this in here. Please remove it!